### ISSUE TITLE
move docker/distribution-library-image to distribution org

### ISSUE THREAD

**Issue by milosgajdos** [distribution/distribution#4124]
This is the repository that's used  to build `registry` official image

**Comment by milosgajdos** [comment]
@thaJeztah would you please take this over? 

I believe there should no longer be any blocker that prevents this from happening. CC: @justincormack

**Comment by thaJeztah** [comment]
I probably don't have permissions to transfer the repository, so yes, @justincormack and maybe @yosifkit or @tianon are probably needed to help moving it.

**Comment by thaJeztah** [comment]
Oh! This was moving it to `distribution`, right? (ISTR there was some talk about moving it to `docker-library`, so if it's `distribution` then _probably_ @yosifkit and @tianon are not needed

I recall @tianon mentioned that having it in `docker-library` would allow them to take care of (part of) maintenance, but I'm a but fuzzy on that part.

**Comment by milosgajdos** [comment]
> I recall @tianon mentioned that having it in docker-library would allow them to take care of (part of) maintenance, but I'm a but fuzzy on that part.

Wouldn't that require a lot of coordination between ourselves and @tianon @yosifkit? I'd like ton hear more about how that'd work. Where things stand at the moment is: that repo is owned by Docker where none of the maintainers bar you and @justincormack have write access over so the end users are doomed to come here and complain when there is a release but there is no DOI image available for it becase we're having to chase Docker folks. Not ideal.

**Comment by thaJeztah** [comment]
> Wouldn't that require a lot of coordination between ourselves and @tianon @yosifkit? 

Maybe! I pinged them on Slack as well (Tianon is still on leave, but I think back "soon"). I agree that it probably should be moved "somewhere" but hopefully they can fill be in on details (pros/cons and options for each)

**Comment by thaJeztah** [comment]
Maybe this was around automation; I know that for (for example) the `golang` image, they have automation running, so whenever a release happens of Go, it automatically updates the images, and it allows them to update the base-image (alpine versions), so _perhaps_ moving it there indeed means "less maintenance" (or "none at all") for the distribution maintainers.

**Comment by milosgajdos** [comment]
Curious to hear more. I'm pretty sure we could hack up something that could even open a PR in library repo on every release. But as I said, this is a blocker at the moment so I'd like to avoid this dragging for too long, hence IMHO, the ideal  _first_ step forward would be moving the existing repo under this org 🤷‍♂️

**Comment by tianon** [comment]
Yep, I'm still out, but this is an easy one from as much as I/we can help - we're happy to take the repo in docker-library and automate version bumps, but we do definitely prefer it be upstream-maintained, and that the existing repository continuity is preserved (issues, commit history), so transferring to distribution makes sense IMO (however huge caveat I can't make that call nor do I have GH permissions to initiate such a thing).

When considering automated PRs (once the repo home is settled), you'll want to review https://github.com/docker-library/faq#can-i-use-a-bot-to-make-my-image-update-prs :bow: :heart:

**Comment by thaJeztah** [comment]
Thanks @tianon!

My train of thought was "where to make the split", and what to consider a release; I think various (most) official images build from source, so was considering if "distribution" maintainers should be dealing with the source, and official repo deal with "build from source if a tag / release is created on GitHub". And the DOI side could handle "we're gonna be updating all images to use `alpine X`".

But I'm not sure what's most common, and what's preferred from both parties (what level of control each wants over the image).

@milosgajdos is there a timeframe you're targeting to complete this work? (Wondering if a (zoom) call would work once Tianon is back to work)

**Comment by milosgajdos** [comment]
> @milosgajdos is there a timeframe you're targeting to complete this work? (Wondering if a (zoom) call would work once Tianon is back to work)

There is no timeframe for this or rather there is one = as soon as the milestone issues are cleared. But I understand that saying this might just completely deprioritise this issue in the minds of people who can actually get this done.

Either way, we will be pushing the images into https://hub.docker.com/r/distribution/distribution and to GHCR; we would love if the end users had access to the releases via the official images, too, but if it becomes an issue that drags for weeks then we'll just point the users to the relevant repo that owns the official image.

**Comment by thaJeztah** [comment]
Yup; let's see how fast we can come up with a plan / best location / move, and set that in motion.

I had a brief chat with @justincormack on Slack, and from Docker's perspective, there's no objections against moving the repository, and to hand over maintenance to distribution and/or official images maintainers.

**Comment by whalelines** [comment]
According to the [README for the `registry` DOI](https://hub.docker.com/_/registry), it is the official image for [this GitHub repo](https://github.com/distribution/distribution). If that is true, then it seems everything is already in place.

Note: I was unable to find any docker/distribution image on Docker Hub.

**Comment by whalelines** [comment]
I should have also noted that Docker Hub does not support redirects, so there is no way to "move" a repository from one namespace to another. Moving is a manual process that involves pulling all the images from the source namespace and pushing them to the destination namespace and then, optionally, deleting them from the source and posting a notice in the source README. The optional step is not recommended since it will break people's builds.

**Comment by thaJeztah** [comment]
> According to the [README for the registry DOI](https://hub.docker.com/_/registry), it is the official image for [this GitHub repo](https://github.com/distribution/distribution). If that is true, then it seems everything is already in place.

Yes, the situation was a bit weird; it was the official image for "docker/distribution", which of course got donated to the CNCF (now distribution/distribution), but ownership of the official image repository (both on Docker Hub, and on GitHub (https://github.com/docker/distribution-library-image)) remained at Docker.

I _think_ there was some motivation behind that at the time (I recall I asked about it, but I forgot the details). In either case, that's no longer the case, so we should be good to go to set things in motion.

And when we conclude the technicalities (where to move, and who to (help) maintain), I can try finding a GitHub admin/owner for the docker org to help transferring the GitHub repository.

**Comment by milosgajdos** [comment]
> Note: I was unable to find any docker/distribution image on Docker Hub.

There is no such thing. Please see the link I've explicitly mentioned in one of my comments: https://hub.docker.com/r/distribution/distribution 

> According to the [README for the registry DOI](https://hub.docker.com/_/registry), it is the official image for [this GitHub repo](https://github.com/distribution/distribution). If that is true, then it seems everything is already in place.

It isn't. No. Docker donated this repository to CNCF but the base repo used to build the official image remained under docker GH org. It's all discussed earlier in this thread.

**Comment by tianon** [comment]
Indeed!

To reiterate, https://hub.docker.com/_/registry is a packaging of the server component of the open source project at https://github.com/distribution/distribution (which was previously https://github.com/docker/distribution and a Docker Inc project, but was donated to the CNCF), so  _is_ the official "upstream" for https://hub.docker.com/_/registry and as such should ideally be maintaining it (similar to https://hub.docker.com/r/distribution/distribution, which differs from the DOI image mainly in that it _also_ contains "edge" / development builds).

So to summarize, my suggestion here would be to further echo what Milos is suggesting; that is:

1. transfer https://github.com/docker/distribution-library-image to https://github.com/distribution  
   (need someone with transfer capability on https://github.com/docker for that)

1. update https://github.com/docker-library/official-images/blob/master/library/registry and https://github.com/docker-library/docs/tree/master/registry (via PRs) to reflect that transfer in location and maintainership appropriately

**Comment by thaJeztah** [comment]
@justincormack are you able to transfer the repo? (do you need someone on the `distribution` side, or do you have all the permissions needed?

**Comment by whalelines** [comment]
@milosgajdos, sorry for the confusion around which repo (GitHub vs. Docker Hub) move was being requested. Please accept my apology.

I'll ping internal folks to try to get a transfer of ownership request initiated for https://github.com/docker/distribution-library-image to the https://github.com/distribution org (@tianon's [item 1](https://github.com/distribution/distribution/issues/4124#issuecomment-1780047781)).

Once that is done, perhaps @milosgajdos you can create the PRs for item 2?

**Comment by milosgajdos** [comment]
> Once that is done, perhaps @milosgajdos you can create the PRs for item 2?

Yes, we're happy to do that indeed.

**Comment by whalelines** [comment]
The [GitHub docs on transferring repositories](https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository) are not entirely clear on this, but it seems when transferring a repo from one org to another, the person doing it needs to have owner-like permissions in both orgs. @justincormack is going to take it from here.

**Comment by milosgajdos** [comment]
@thaJeztah what is the status of this? Is that still with @justincormack? Can you ping him, please.

**Comment by thaJeztah** [comment]
Yes, let me give him a nudge again

**Comment by thaJeztah** [comment]
He's attending KubeCon, and told me he would be looking at this after that.

**Comment by milosgajdos** [comment]
Any progress with this @thaJeztah

**Comment by fyockm** [comment]
Am I correct in understanding that this is the last remaining task in the [milestones](https://github.com/distribution/distribution/milestone/22) for the v3.0.0 Release? if so, perhaps it's time for another nudge?

**Comment by justincormack** [comment]
moved, sorry for delay
